### DISCUSSION TITLE
Query returns null for all referenced rows

### DISCUSSION THREAD

**Discussion by jameshfisher** [supabase/supabase#20624] [Questions] [answered]
I'm making this API request:

```
https://MY_DB.supabase.co/rest/v1/game?select=id,club(slug)&apikey=MY_API_KEY
```

I get this response:

```
[{"id":"b4771c4b-e8a6-425d-8720-f53a9f93af77","club":null}, 
 {"id":"df5d0c95-47bc-4db3-b29b-eed7d4b286c1","club":null}, 
 {"id":"c9d1b696-5a44-4097-89e1-da2e9b49b7d5","club":null}]
```

All returned rows have `"club":null`, but this is unexpected. Each of those rows has `club_id=1`, which references the `id` column of the `club` table. The response should therefore be:

```
[{"id":"b4771c4b-e8a6-425d-8720-f53a9f93af77","club":{"slug":"club1"}}, 
 {"id":"df5d0c95-47bc-4db3-b29b-eed7d4b286c1","club":{"slug":"club1"}}, 
 {"id":"c9d1b696-5a44-4097-89e1-da2e9b49b7d5","club":{"slug":"club1"}}]
```

Here is my database:

```
postgres=> select id, club_id from game;
                  id                  | club_id
--------------------------------------+---------
 b4771c4b-e8a6-425d-8720-f53a9f93af77 |       1
 df5d0c95-47bc-4db3-b29b-eed7d4b286c1 |       1
 c9d1b696-5a44-4097-89e1-da2e9b49b7d5 |       1
(3 rows)

postgres=> select id, slug from club;
 id |     slug
----+---------------
  1 | club1
...
(4 rows)

postgres=> \d game
...
Foreign-key constraints:
    "game_club_id_fkey" FOREIGN KEY (club_id) REFERENCES club(id)
...
```

My understanding/guess is that the Supabase query `game?select=id,club(slug)` will magically detect `game_club_id_fkey` constraint. Using this, it should generate a SQL query like:

```
postgres=> select g.id, c.slug from game g join club c on g.club_id = c.id;
                  id                  |   slug
--------------------------------------+-----------
 b4771c4b-e8a6-425d-8720-f53a9f93af77 | club1
 df5d0c95-47bc-4db3-b29b-eed7d4b286c1 | club1
 c9d1b696-5a44-4097-89e1-da2e9b49b7d5 | club1
(3 rows)
```

Clearly this is not what is happening. But how do I debug this?

I've tried to find the specification for the query language `game?select=id,club(slug)`. But where are those docs? All I can see is https://supabase.com/docs/reference/javascript/select, which is basically just a list of examples with no commentary.

I've tried to find logs, which might show the underlying SQL query. My understanding is that this is PostgREST, so I tried https://supabase.com/dashboard/project/MY_DB/logs/postgrest-logs, but this shows nothing ("No results").

And/or, I'm looking for how to debug this. I'm completely stuck.

**Comment by jameshfisher** [comment]
It's particularly confusing that the docs call this example "Query referenced tables". It presumably should be called "query referencing tables", because the `cities` reference the `countries`, rather than the other way around:

<img width="1312" alt="Screenshot 2024-01-22 at 9 40 48" src="https://github.com/supabase/supabase/assets/166966/2301d5ce-a43e-4d7d-943f-beaeed5747ca">

**Comment by GaryAustin1** [comment, marked as answer]
Without reading thru your code (have not had coffee yet)... the most common issue with getting null is RLS on the 2nd table not being met.
If the column is a unique column you will get null, if a non-unique column an empty array.

**Comment by jameshfisher** [comment]
Thanks @GaryAustin1! I was just posting this as you replied: **The problem was row-level security. The policies for the club table were not allowing access. I fixed this by adding a new policy allowing anonymous read access to the club table.**

That said, the underlying "debuggability" problem is still unsolved ... what I really want is to see and run the SQL that's presumably behind the scenes of this API request. Is there an easy way to do that?

**Reply by GaryAustin1** [reply]
You can debug with the SQL editor and impersonation not available to act as anon or authenticated users.
There is no Postgres ability to monitor RLS with logs or to error versus just silently ignore columns you don't have access to.

**Reply by jameshfisher** [reply]
Thanks Gary, that makes sense, but my main problem was, "what SQL query is this API request running?" I think if I had access to that SQL, plus the SQL editor and impersonation, I would have realized what the problem was much more quickly.

I just looked at https://status.supabase.com/ and apparently logs are down. Perhaps that's why I didn't have access to this!

**Reply by GaryAustin1** [reply]
There is a way to get the SQL in the logs.  I don't have it off the top of my head what setting you have to do.  BUT it is pretty useless info as the actual query looks nothing like a basic SQL query doing the operation.  PostgREST has to do prepared statements, and also deal with caches for foreign keys and functions.  I looked at it a couple of times and it was a waste of time.  
Basically your query is doing a join using the foreign key reference it finds in its cache.  If you use !inner flag then it is doing an inner join.   This is the best reference on PostgREST the REST API Supabase uses:  https://postgrest.org/en/stable/references/api.html

**Reply by steve-chavez** [reply]
>  "what SQL query is this API request running?"
> BUT it is pretty useless info as the actual query looks nothing like a basic SQL query doing the operation
> PostgREST has to do prepared statements, and also deal with caches for foreign keys and functions. I looked at it a couple of times and it was a waste of time.

There are plans on PostgREST to obtain the generated SQL on a per-request basis but as Gary mentions you'll not be able to just run it because it will be a prepared statement with parametrized values.

The thing is that RLS is an advanced PostgreSQL feature that Supabase makes too easy to use, it affects every query silently. I'm not sure how to help newcomers who think that there might be something wrong with the generated query instead of thinking something is off with their schema definition. I welcome any suggestions.

Related upstream discussion https://github.com/PostgREST/postgrest-docs/issues/749.
